### PR TITLE
:white_check_mark: test: Add Tests

### PR DESCRIPTION
**Pull request by YashGaykar0309**
## Description

Added tests for following resources.
1. VM
2. Security Group
3. Net and SubNet
4. Snapshot
5. Volume
6. LoadBalancer
7. EIM User

Updates old examples to sync with new tests.

Also updated the examples README with clearer instructions for configuring credentials and running tests.

## Type of Change

Please check the relevant option(s):

- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 🧹 Code cleanup or refactor
- [x] 📝 Documentation update
- [ ] 🔧 Build or CI-related change
- [ ] 🔒 Security fix
- [x] Other (specify): Test addition

## How Has This Been Tested?

Please describe the test strategy:

- [ ] Manual testing
- [ ] Unit tests
- [x] Integration tests
- [ ] Not tested yet

Commands used :

```bash
# Example
go test ./examples
````

## Checklist

* [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
* [x] I have added tests or explained why they are not needed
* [x] I have updated relevant documentation (README, examples, etc.)
* [x] My changes follow the [Conventional Commits](https://www.conventionalcommits.org/) specification
* [x] My commits include appropriate [Gitmoji](https://gitmoji.dev/)

### CODE REVIEW

File: examples/security_group_test.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+// 2. Add an inbound SSH rule.
+// 3. Read the security group and validate the rule.
+// 4. Delete the security group rule.
+// 5. Delete the security group.
+func TestSecurityGroup(t *testing.T) {
+	ctx := t.Context()
+
+	userProfile, err := profile.New()
+	if err != nil {
+		panic(err)
```
Comment (jfbus) [inline examples/security_group_test.go]: do not use panic to do assertions in tests.

You may use testify:

`require.NoError(t, err)`

or stdlib:

`t.Fatal`

File: examples/security_group_test.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	client, err := osc.NewClient(userProfile, options.WithLogging(&testingLogger{t}))
+	if err != nil {
+		panic(err)
+	}
+
+	securityGroupName := fmt.Sprintf("OscGoSdkTest-%s", RandomString(10))
```
Comment (jfbus) [inline examples/security_group_test.go]: string concatenation is more efficient:

`securityGroupName := "OscGoSdkTest-"+RandomString(10)`

File: examples/security_group_test.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		SecurityGroupName: securityGroupName,
+		Description:       "Test security group lifecycle",
+	}
+
+	createResp, err := client.CreateSecurityGroup(ctx, createReq)
+	if err != nil {
+		panic(err)
+	}
+
+	if createResp.SecurityGroup == nil {
```
Comment (jfbus) [inline examples/security_group_test.go]: `require.NotNil(t, createResp.SecurityGroup)` (or stdlib)

File: examples/security_group_test.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		panic("SecurityGroup is nil in response")
+	}
+
+	securityGroupID := createResp.SecurityGroup.SecurityGroupId
+	if securityGroupID == "" {
```
Comment (jfbus) [inline examples/security_group_test.go]: `require.NotEmpty(t, createResp.SecurityGroup.SecurityGroupId)` (or stdlib)

File: examples/security_group_test.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		if rule.FromPortRange != fromPort {
+			continue
+		}
+		if rule.ToPortRange != toPort {
+			continue
+		}
+		if rule.IpProtocol != tcp {
+			continue
+		}
+		if slices.Contains(rule.IpRanges, ipRange) {
```
Comment (jfbus) [inline examples/security_group_test.go]:
```
if slices.Contains(rule.IpRanges, ipRange) {
  foundSSHRule = true
  break
}
```

**Review by jfbus — CHANGES_REQUESTED** [review]

File: examples/eim_user_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			UserIds: &[]string{userID},
+		},
+	})
+	require.NoError(t, err)
+	require.NotNil(t, readResp.Users)
+	require.Len(t, *readResp.Users, 1)
+
+	user := (*readResp.Users)[0]
+	require.NotNil(t, user.UserId)
+	require.Equal(t, userID, *user.UserId)
```
Comment (jfbus) [inline examples/eim_user_test.go]: You should not use require everywhere

require is only required when the following values tested are invalid : the call returned an error and you need to test the result, a pointer is nil and you need to test its value or a slice has not the right len and you need to test its contents

use assert.X

**Review by jfbus — CHANGES_REQUESTED** [review]

File: examples/eim_user_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	createResp, err := client.CreateUser(ctx, osc.CreateUserRequest{
+		Path:      &userPath,
+		UserEmail: &userEmail,
+		UserName:  userName,
+	})
+	require.NoError(t, err)
+	require.NotNil(t, createResp.User)
+	require.NotNil(t, createResp.User.UserId)
+	assert.NotNil(t, createResp.User.UserName)
```
Comment (jfbus) [inline examples/eim_user_test.go]: you might want to also test the value

File: examples/eim_user_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+	userID := *createResp.User.UserId
+	assert.NotEmpty(t, userID)
+
+	t.Logf("Created EIM user: %s", userID)
+
+	defer func() {
```
Comment (jfbus) [inline examples/eim_user_test.go]: you should move the defer just after `require.NoError(t, err)`

if `require.NotNil(t, createResp.User)` fails, the defer will not be executed and you'll be left with a user...

File: examples/load_balancer_backend_test.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+	foundBackend := false
+	for _, backendVMID := range lb.BackendVmIds {
+		if backendVMID == vmID {
+			foundBackend = true
+			break
+		}
+	}
+	require.True(t, foundBackend, "expected backend VM %s on load balancer %s", vmID, lbName)
+
+	var healthEntry *osc.BackendVmHealth
```
Comment (jfbus) [inline examples/load_balancer_backend_test.go]: you shoud name your loop

```
LOOPHEALTH:
for range 18 {
```

File: examples/load_balancer_backend_test.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+		healthResp, err := client.ReadVmsHealth(ctx, osc.ReadVmsHealthRequest{
+			LoadBalancerName: lbName,
+			BackendVmIds:     &[]string{vmID},
+		})
+		require.NoError(t, err)
+		if healthResp.BackendVmHealth != nil {
+			for _, entry := range *healthResp.BackendVmHealth {
+				if entry.VmId != nil && *entry.VmId == vmID {
+					healthEntry = &entry
+					break
```
Comment (jfbus) [inline examples/load_balancer_backend_test.go]: break LOOPHEALTH

File: examples/load_balancer_backend_test.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+				}
+			}
+		}
+		if healthEntry != nil {
```
Comment (jfbus) [inline examples/load_balancer_backend_test.go]: no need for this if using break LOOPHEALTH

File: examples/load_balancer_backend_test.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+			break
+		}
+
+		time.Sleep(10 * time.Second)
+	}
+	require.NotNil(t, healthEntry, "expected backend health entry for VM %s", vmID)
+
+	t.Logf("Backend VM %s registered to load balancer %s with health state %v", vmID, lbName, healthEntry.State)
```
Comment (jfbus) [inline examples/load_balancer_backend_test.go]: shouldn't you test deletion ?

File: examples/load_balancer_backend_test.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+
+	var healthEntry *osc.BackendVmHealth
+	for range 18 {
+		healthResp, err := client.ReadVmsHealth(ctx, osc.ReadVmsHealthRequest{
+			LoadBalancerName: lbName,
+			BackendVmIds:     &[]string{vmID},
+		})
+		require.NoError(t, err)
+		if healthResp.BackendVmHealth != nil {
+			for _, entry := range *healthResp.BackendVmHealth {
```
Comment (jfbus) [inline examples/load_balancer_backend_test.go]: as you are not really using healthEntry, you might have used `slices.ContainsFunc` to check that the list contains an entry for your Vm...

(if using slices.ContainsFunc, no need to name the loop)

File: examples/net_subnet_test.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	readResp, err := client.ReadSubnets(ctx, readReq)
+	require.NoError(t, err)
+	require.NotNil(t, readResp.Subnets)
+	require.Len(t, *readResp.Subnets, 1)
+	require.Equal(t, subnetID, (*readResp.Subnets)[0].SubnetId)
+
+	foundSubnetTag := false
+	for _, tag := range (*readResp.Subnets)[0].Tags {
```
Comment (jfbus) [inline examples/net_subnet_test.go]: you can use assert.Contains (tags are comparable as no pointer is used)

note: outside tests, slices.Contains works

File: examples/project_test.go [hunk 15]
```diff
@@ -1,78 +1,69 @@
...
 	project := oks.CreateProjectJSONRequestBody{Name: name}
 	err = mergo.Merge(&project, projectTemplate.Template)
-	if err != nil {
-		panic(err)
-	}
+	require.NoError(t, err)
 
 	var projectID string
 	createProject, err := client.CreateProject(ctx, project)
 	if err != nil {
```
Comment (jfbus) [inline examples/project_test.go]: complex if/else tests are usually better represented by switch/case:
```
switch {
case oks.IsConflict(err):
case err != nil:
default:
}
```

One Go rule is : "try to keep your code at the lowest indentation possible"

File: examples/project_test.go [hunk 16]
```diff
@@ -1,78 +1,69 @@
...
 
 	client, err := oks.NewClient(userProfile, options.WithLogging(&testingLogger{t}))
-	if err != nil {
-		panic(err)
-	}
+	require.NoError(t, err)
 
 	ctx := t.Context()
 
 	_, err = client.GetProject(ctx, "a-non-existing-project")
```
Comment (jfbus) [inline examples/project_test.go]: this is a different test than the next, you should isolate both in t.Run blocks:

```
t.Run("Fetching a non existent project should return a not found error", func()...
```

Comment (jfbus) [reply examples/project_test.go]: This test is a good idea, it would be interesting to have the same for all resources.

File: examples/project_test.go [hunk 17]
```diff
@@ -1,78 +1,69 @@
...
 
 	project := oks.CreateProjectJSONRequestBody{Name: name}
 	err = mergo.Merge(&project, projectTemplate.Template)
-	if err != nil {
-		panic(err)
-	}
+	require.NoError(t, err)
 
 	var projectID string
 	createProject, err := client.CreateProject(ctx, project)
```
Comment (jfbus) [inline examples/project_test.go:44]: As you create OKS resources, you will need to configure frieza to cleanup OKS:

```
            - uses: outscale/frieza-github-actions/frieza-clean@master
              with:
                  access_key: ...
                  secret_key: ...
                  region: ...
                  providers: outscale_oapi,outscale_oks
```

File: examples/security_group_test.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	readResp, err := client.ReadSecurityGroups(ctx, readReq)
+	require.NoError(t, err)
+	require.NotNil(t, readResp.SecurityGroups)
+	require.Len(t, *readResp.SecurityGroups, 1)
+
+	sg := (*readResp.SecurityGroups)[0]
+	foundSSHRule := false
+	for _, rule := range sg.InboundRules {
```
Comment (jfbus) [inline examples/security_group_test.go]: you can also use slices.ContainsFunc, but I'm not sure that it would be a great improvement, I'll let you see

File: examples/snapshot_test.go [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+		require.NoError(t, err)
+		require.NotNil(t, readVolumesResp.Volumes)
+		require.Len(t, *readVolumesResp.Volumes, 1)
+
+		volume := (*readVolumesResp.Volumes)[0]
+		if volume.State == osc.VolumeStateAvailable {
+			break
+		}
+		if volume.State == osc.VolumeStateError {
+			t.Fatalf("volume %s entered error state", volumeID)
```
Comment (jfbus) [inline examples/snapshot_test.go]: you can replace this with a require.NotEqual

File: examples/snapshot_test.go [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+		require.NoError(t, err)
+		require.NotNil(t, readResp.Snapshots)
+		require.Len(t, *readResp.Snapshots, 1)
+
+		snapshot := (*readResp.Snapshots)[0]
+		t.Logf("Snapshot %s state: %s", snapshotID, snapshot.State)
+		if snapshot.State == osc.SnapshotStateCompleted {
+			break
+		}
+		if snapshot.State == osc.SnapshotStateError {
```
Comment (jfbus) [inline examples/snapshot_test.go]: require.NotEqual ?

File: examples/snapshot_test.go [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+	require.Len(t, *readResp.Snapshots, 1)
+
+	snapshot := (*readResp.Snapshots)[0]
+	require.Equal(t, snapshotID, snapshot.SnapshotId)
+	require.Equal(t, volumeID, snapshot.VolumeId)
+	require.NotNil(t, snapshot.Description)
+	require.Equal(t, description, *snapshot.Description)
+	foundSnapshotTag := false
+	require.NotNil(t, snapshot.Tags)
+	for _, tag := range *snapshot.Tags {
```
Comment (jfbus) [inline examples/snapshot_test.go]: assert.Contains

File: examples/vms_test.go [hunk 22]
```diff
@@ -1,30 +1,177 @@
...
+		require.NoError(t, err)
+		require.NotNil(t, readResp.Vms)
+		require.Len(t, *readResp.Vms, 1)
+
+		vm := (*readResp.Vms)[0]
+		t.Logf("VM %s state: %s", vmID, vm.State)
+		if vm.State == osc.VmStateRunning {
+			break
+		}
+		if vm.State == osc.VmStateTerminated || vm.State == osc.VmStateShuttingDown {
```
Comment (jfbus) [inline examples/vms_test.go]: require.NotEqual

File: examples/vms_test.go [hunk 23]
```diff
@@ -1,30 +1,177 @@
...
+	require.NoError(t, err)
+	require.NotNil(t, readResp.Vms)
+	require.Len(t, *readResp.Vms, 1)
+
+	vm := (*readResp.Vms)[0]
+	require.Equal(t, osc.VmStateRunning, vm.State)
+	require.Equal(t, imageID, vm.ImageId)
+
+	foundTag := false
+	for _, tag := range vm.Tags {
```
Comment (jfbus) [inline examples/vms_test.go]: assert.Contains

**Review by jfbus — CHANGES_REQUESTED** [review]

File: examples/eim_user_test.go (hunk 6 above)
Comment (YashGaykar0309) [reply examples/eim_user_test.go]: Done, Thanks

File: examples/eim_user_test.go (hunk 8 above)
Comment (YashGaykar0309) [reply examples/eim_user_test.go]: I have added a flag-based defer for all resources. So even if resource deletion is skipped explicitly, the deferred cleanup will still ensure the resources are cleared.

https://github.com/outscale/osc-sdk-go/pull/1064/changes/04dc17d77eeed514d2ddd1021c4a5a50be4aad8c

**Review by jobs62 — APPROVED** [review]
